### ISSUE TITLE
Making release process fully automated

### ISSUE THREAD

**Issue by mikrosk** [atarijookie/ce-atari#113]
Just a follow-up from #108...

- generate ce_conf.zip by zip in commandline and make
- generate ce_conf.st - by mtools and make

**Comment by miniupnp** [comment]
I dont like the idea of having binary files stored in git.
But well as ATARI progs have to be build on PC and RPi on RPi we may be forced to do so.

@atarijookie : I want to make a Makefile that build all ATARI progs and generate the configdrive (on PC of course). How integrate this with your release process ?

**Comment by atarijookie** [comment]
I'm not sure if we want to automatically build all the atari binaries, as
now I build and release only the one(s) which have been changed and are
considered now in a good shape... but as long as the master branch would be
considered as stable / in good shape, then it could work...

I don't have any plan or idea how to make the whole a single step process,
because as you mentioned - one is compiled on raspbian, one on yocto, and
the rest under windows with cross compiler...

But anything would be better than nothing...

**Comment by tin-nl** [comment]
Just a note: the changes in the makefiles did make cross compiling ce_main under windows more and more difficult (e.g. code generation etc.). It would be very nice to keep cross compiling in mind before introducing more complexity to the makefiles.

**Comment by miniupnp** [comment]
@tin-nl just tell me which changes causes problem. you can open issues like "build failed on cygwin..." 
I will try to adapt things
I'm sorry, I'm too used to the comfort of POSIX environments :) but isn't MS windows supposed to include a GNU/Linux subsystem soon ?

**Comment by tin-nl** [comment]
@miniupnp: no need to be sorry :) Normally I'd build under linux/OSX too, but in this case I'm building under windows deliberately - to make sure win cross compiling is possible and keeps being possible. 
It's not that it's not working per se, it just gets more difficult (last by introducing code generation via xxd).
As said before: it would be nice to keep cross compiling in mind before introducing more complexity into the makefiles.
I'll compile some things into a dedicated ticket re current state of cross compilation.

**Comment by miniupnp** [comment]
@atarijookie that kind of simple stuff can help ? https://github.com/miniupnp/ce-atari/tree/global_makefile

**Comment by atarijookie** [comment]
@miniupnp Yes, something like that, but I would also add after the build something like:
- create some output (release / update dir)
- copy the files there
- ZIP them 
- then sick the mtools on them

The current structure, which would be lovely to achieve using that makefile, can be found in the current app update package:
joo.kie.sk/cosmosex/update/app.zip

This is a good start :+1:

**Comment by miniupnp** [comment]
ok. I'm away for vacation, I'll work on this next week ;)

**Comment by miniupnp** [comment]
I'm starting to work on this.
What is fdd_test.st ?

**Comment by miniupnp** [comment]
check my `update_package` branch. just "make update" to build the app.zip file !
it supposes the atari progs have been build previously (make atari_progs)
and the cosmosex_* RPi apps too

only thing to fix is fdd_test.st.

**Comment by miniupnp** [comment]
@atarijookie also tell me how to make package for release ;)

**Comment by atarijookie** [comment]
> only thing to fix is fdd_test.st.

This is how it is generated - it was used on the first start of new ce_main_app. It's a simple image which is used for floppy testing, when using tests/ce_tstfd,tos
https://github.com/atarijookie/ce-atari/blob/master/ce_main_app/update.cpp#L513-L548

> also tell me how to make package for release ;)

I'm not sure what you mean by that one - making that app.zip is all that is needed, then I just upload it to web (through ftp), update version and checksum in the .csv and that's all. The cosmosex_* binaries are just striped before adding them to app.zip

**Comment by miniupnp** [comment]
I suppose you are building  cosmosex_* on real devices. I'm copying them from ce_main_app/ directory, it is easy to change that part to suit your needs.
It would be great to have cross compiling toolchain so everything is built with a single make command but that's a bigger thing :)

about the fdd_test.st, I have gzipped it and I think that is ok for now :)
I was wondering : isn't there some issue because the bootsector is not standard ?

**Comment by miniupnp** [comment]
@atarijookie I have rebased my branch [update_package](https://github.com/miniupnp/ce-atari/tree/update_package) so it will be easier for you to merge once you tested it

**Comment by atarijookie** [comment]
> I suppose you are building cosmosex_* on real devices. I'm copying them from ce_main_app/ directory, it is easy to change that part to suit your needs.

Yes, and ok.

> It would be great to have cross compiling toolchain so everything is built with a single make command

Yes, it would, but even mr. crosscompile (MiKRO) didn't get this running as far as I know. 

> I was wondering : isn't there some issue because the bootsector is not standard ?

Usually no, because almost no one is using it, and when you use it, you can load it during the test, so you won't let TOS try to boot from it and cause trouble. 

> ...so it will be easier for you to merge once you tested it

If it doesn't break the old way of doing things (manual make of ce_main_app, manual make of ST apps), then I suggest that you just merge it right away :)

**Comment by miniupnp** [comment]
> f it doesn't break the old way of doing things (manual make of ce_main_app, manual make of ST apps), then I suggest that you just merge it right away :)

The only thing I "broke" is I removed the files from git in ce_main_app/configfile

**Comment by atarijookie** [comment]
> The only thing I "broke" is I removed the files from git in ce_main_app/configfile

Then you didn't break anything, and please just merge it :)

**Comment by miniupnp** [comment]
done.
I'm closing. @mikrosk can re-open once cross-compiling for RPi is available ;)

**Comment by atarijookie** [comment]
Currently it seems it doesn't work under Windows + cygwin:
- can't make bootsectors, as make can't compile native app on my cygwin and thus convert those PRGs to .BS
- can't compile Sting test server app under cygwin
- bad argument to cp when doing 'make update', it then stops the make:
cp -p ../ce_dd_bootsectors/ce_dd.bs configdrive/drivers/ce_dd.bs
cp: preserving permissions for ‘configdrive/drivers/ce_dd.bs’: Invalid argument

- MTOOLS not present on cygwin, so it can't create ce_conf.st and fails there

It seems it would be too much effort to fix it under Windows...

**Comment by mikrosk** [comment]
@miniupnp said I should reopen it once cross-compiling for the RPi is available and it is, for quite some time: https://github.com/atarijookie/ce-atari/wiki/How-to-setup-an-RPi-cross-compiler ;)

**Comment by miniupnp** [comment]
`$ make ce_main_app CC=arm-linux-gnueabihf-gcc CXX=arm-linux-gnueabihf-g++ DISTRO=raspbian`
Does work 

@atarijookie I guess you can install cygwin mtools. can you give error output of make for boot sectors and sting test server ?

**Comment by miniupnp** [comment]
@atarijookie ping ? :)
I fixed the "cp -p" issue, but can you tell me which errors you have when trying to build boot sectors and sting test server with cygwin ?

**Comment by atarijookie** [comment]
@miniupnp 
I'm currently hibernating (not doing anything), but there's some effort going on to get floppy write done, so if people will help me a bit with tools, I will resurrect myself and I'll let you know about this...

**Comment by mikrosk** [comment]
Additionally, maybe we could finally tune the CE executable for performance? Right now we release unoptimised binaries. This makes debugging easier but I don't think there have been many reports where a core dump was the only option how to investigate.

Nice table how different RPi models can be optimised: https://gist.github.com/fm4dd/c663217935dc17f0fc73c9c81b0aa845

**Comment by miniupnp** [comment]
the most important thing for performance is to remove all polling in the ce_main_app...
I have a branch with GPIO interrupt support

**Comment by mikrosk** [comment]
Anyways - I have updated the wiki with my latest findings about cross-compilers: we can use Ubuntu's `g++-arm-linux-gnueabihf` package + cross compile the four (bcm2835, zlib, openssl, curl) packages during Travis `install` phase.

Commit https://github.com/atarijookie/ce-atari/commit/2887b43c5b1d154d26f1d87bbb40be40fd2b6546 makes the needed flags explicit so really, all we need is some cleanup and support for make CROSS=yes. :)

EDIT: actually, maybe we need only two packages (bcm2835 and curl), since we are not going to use them anyway. So it doesn't matter that curl is not going to have zlib/openssl support.

EDIT 2: Wow, now I see it has been done exactly like that in .travis.yml, cool. So we really just need to create snapshots on bintray or so.
